helpers/manifest-parser.js

```diff
@@ -0,0 +1,325 @@
+  let display = parseString(jsonInput.display, true);
+
+  if (display.value && ALLOWED_DISPLAY_VALUES.indexOf(display.value.toLowerCase()) === -1) {
+    display.value = undefined;
+    display.warning = 'ERROR: \'display\' has an invalid value, will be ignored.';
+  }
+
+  return display;
+}
+
+function parseOrientation(jsonInput) {
+  let orientation = parseString(jsonInput.orientation, true);
+
+  if (orientation.value &&
+      ALLOWED_ORIENTATION_VALUES.indexOf(orientation.value.toLowerCase()) === -1) {
+    orientation.value = undefined;
+    orientation.warning = 'ERROR: \'orientation\' has an invalid value, will be ignored.';
+  }
+
+  return orientation;
+}
+
+function parseIcon(raw) {
+  // TODO: pass manifest url as base.
+  let src = parseURL(raw.src);
+  let type = parseString(raw.type, true);
+
+  let density = {
+    raw: raw.density,
+    value: 1,
+    warning: undefined
+  };
+  if (density.raw !== undefined) {
+    density.value = parseFloat(density.raw);
+    if (isNaN(density.value) || !isFinite(density.value) || density.value <= 0) {
+      density.value = 1;
+      density.warning = 'ERROR: icon density cannot be NaN, +∞, or less than or equal to +0.';
+    }
+  }
+
+  let sizes = parseString(raw.sizes);
+  if (sizes.value !== undefined) {
+    let set = new Set();
+    sizes.value.split(/\s/).forEach(size => set.add(size.toLowerCase()));
+
+    sizes.value = set.size > 0 ? set : undefined;
+  }
+
+  return {
+    raw,
+    value: {
+      src,
+      type,
+      density,
+      sizes
+    },
+    warning: undefined
+  };
+}
+
+function parseIcons(jsonInput) {
+  const raw = jsonInput.icons;
+  let value;
+
+  if (raw === undefined) {
+    return {
+      raw,
+      value,
+      warning: undefined
+    };
+  }
+
+  if (!Array.isArray(raw)) {
+    return {
+      raw,
+      value,
+      warning: 'ERROR: \'icons\' expected to be an array but is not.'
+    };
+  }
+
+  // TODO(bckenny): spec says to skip icons missing `src`. Warn instead?
+  value = raw.filter(icon => !!icon.src).map(parseIcon);
+
+  return {
+    raw,
+    value,
+    warning: undefined
+  };
+}
+
+function parseApplication(raw) {
+  let platform = parseString(raw.platform, true);
+  let id = parseString(raw.id, true);
+  // TODO: pass manfiest url as base.
+  let url = parseURL(raw.url);
+
+  return {
+    raw,
+    value: {
+      platform,
+      id,
+      url
+    },
+    warning: undefined
+  };
+}
+
+function parseRelatedApplications(jsonInput) {
+  const raw = jsonInput.related_applications;
+  let value;
+
+  if (raw === undefined) {
+    return {
+      raw,
+      value,
+      warning: undefined
+    };
+  }
+
+  if (!Array.isArray(raw)) {
+    return {
+      raw,
+      value,
+      warning: 'ERROR: \'related_applications\' expected to be an array but is not.'
+    };
+  }
+
+  // TODO(bckenny): spec says to skip apps missing `platform`. Warn instead?
+  value = raw.filter(application => !!application.platform).map(parseApplication);
```

Comment:
the spec treats applications without a `platform` the same as above with icons without a `src`